### DISCUSSION TITLE
Dynamic Breaks

### DISCUSSION THREAD

**Discussion by rahulj2409** [google/or-tools#2693] [Routing (and legacy CP) questions]
Is there a way to enforce this constraint:

"Drive for X hours, then you MUST take a break for Y hours" in the Vehicle Routing Problem model?

Thanks!

**Comment by Mizux** [comment]
please take a look at https://github.com/google/or-tools/blob/stable/ortools/constraint_solver/samples/vrp_breaks.py

Basically you can use it so solver will place break time inside slackVar (i.e. slackVar range must be large enough)

**Reply by rahulj2409** [reply]
I'm confused, where/what should I be looking at here? How does this help create a dynamic break as well?

Thanks!

**Reply by Mizux** [reply]
by default break need service time and thus are already dynamic on the transit time...

**Reply by rahulj2409** [reply]
I'm confused on the purpose of service times in breaks. As far as I know, a break should a beginning time IE: begin at 9 AM and should last 60 minutes. How is there a service time associated with that/what is the role of a service time there.

Furthermore, I am still confused on how this answers my question of having dynamic breaks IE
"Vehicle 1 can only drive in 10 hour chunks, then must take a 3 hour break, then can restart again"

**Reply by Mizux** [reply]
When you create a dimension, the transit callback will return the travel time + service time (ed also call pre-travel time).

The solver won't place the break if it overlaps with the service time, i.e. vehicle arrive at location, perform its task then can take a break if needed.

So you need to pass the service time to the solver since the transit callback merge both.
 In the sample vrp_break.py (with service time of 15min) you can see:
 ```
Break for vehicle 0: Start(56) Duration(10)
...
0 Time(0) -> 7 Time(9) -> 1 Time(41)
```
So the break starts at 41 (arrival at location 1) +15 (service time) = 56min

Does your 10hours chunk is reset after visit ? if yes, simply add the break(s) in the transit callback return value.

Otherwise you can also use an extra travel only dimension (i.e. with service time 0) so your break will only depends on the driving hours. then add constraint to link your travel dimension slackVar  with the slack var of you time dimension...

**Reply by rahulj2409** [reply]
I just want the simple functionality that after driving more than 10 hours, it takes a break for x amount of time before resuming work again. How do I do this, are there any examples regarding how to do this. I feel like I'm more confused here.
